Expand settlement documentation (#329)
digital-asset/daml-finance: docs/source/concepts/settlement.rst

:ref:`Settlement <settlement>` refers to the execution of holding transfers originating from
a financial transaction.

The library provides facilities to execute these transfers atomically (i.e., within the same Daml transaction).

> Interfaces are defined in the ``Daml.Finance.Interface.Settlement`` package, whereas implementations are is provided in ``Daml.Finance.Settlement``.

Reviewer: are is -> are

In this page we illustrate the settlement workflow with the help
of an example FX transaction, where Alice transfers a EUR-denominated holding
to Bob, in exchange for a USD-denominated holding of the same amount.

We then delve into the details of each of the settlement components.

Workflow
********

Our initial state looks as follows:

> * Alice owns a holding on an `EUR` instrument, for an amount of `1000`

Reviewer: a `EUR`

* Bob owns a holding on a `USD` instrument, for an amount of `1000`

> These holding are generally held at different custodians.

Reviewer: holdings

.. image:: ../images/settlement_initial_state.png
   :alt: Alice owns a EUR holding, Bob owns a USD holding.

Instruct
========

Alice and Bob want to exchange their holdings and agree to enter into the transaction by being signatories on a transaction contract.
Settlement can then be instructed: this generates

#. an ``Instruction`` to transfer 1000 EUR from Alice to Bob
#. an ``Instruction`` to transfer 1000 USD from Bob to Alice
#. a ``Batch`` contract which is used to execute the above Instructions.

.. image:: ../images/settlement_instructed.png
   :alt: Settlement is instructed.

Each instruction defines, for a step of the settlement chain, who is the sender, who is the receiver and what should be transferred (instrument and amount).

Allocate and Approve
====================

In order to perform the transfer, we now need to specify for each ``Instruction``

#. which holding should be used
#. to which account it should be transferred.

> Alice and Bob ``allocate`` the instruction where they are, respectively, the sender by pledging their holding.

Reviewer: Maybe more clear if you describe it only for Alice (and in the next sentence say something like "Bob does the same").

.. image:: ../images/settlement_allocated.png
   :alt: Settlement is allocated.

> Each receiver can then specify to which account the holding should be sent to by ``approving`` the corresponding instruction.

Reviewer: sent to -> sent

.. image:: ../images/settlement_allocated_approved.png
   :alt: Settlement is allocated and approved.

Execute
=======

Once instructions are allocated and approved, a Settler party uses the ``Batch`` contract to ``execute`` them and finalize settlement in one atomic transaction.

.. image:: ../images/settlement_executed.png
   :alt: Settlement is allocated and approved.

> The instructions and batch are archived following a successful execution.

Reviewer: the batch

Remarks
=======

There are some assumptions that need to be verified in order for the transaction above to settle successfully in practice. On the first instruction

- Bob needs to have an account at the custodian where Alice's holding is held
- Alice's holding needs to be :ref:`Transferable <type-daml-finance-interface-holding-transferable-transferable-24986>`
- The transfer must be fully authorized (TODO : add a link to a page where we describe the role of account instructors and approvers, probably somewhere in the asset model)

> 

Reviewer: Do you want to fix the TODO in this or in a separate PR?

Author: Yes, I raised https://github.com/digital-asset/daml-finance/issues/330 for it yesterday. Will remove the TODO from the code

Let us assume, for instance, that Alice's EUR holding in the example above is held at Bank A, whereas Bob has a EUR account at Bank B. Bank A and Bank B both have accounts at the Central Bank.

.. image:: ../images/settlement_hierarchy.png
   :alt: Hierarchical account structure. Alice has an account at Bank A. Bob has an account at Bank B. Bank A and Bank B have an account at the Central Bank.

In this case, a direct holding transfer from Alice and Bob cannot generally be instructed. The original step needs to be broken down by the Settlement Factory into three separate ``Instructions``:

- **1A**: Alice sends 1000 EUR (held at Bank A) back to Bank A
- **1B**: Bank A sends 1000 EUR (held at the Central Bank) to Bank B.
- **1C**: Bank B credits 1000 EUR to Bob's account (held at Bank B)

.. image:: ../images/settlement_hierarchy_instructed.png
   :alt: Instructions for intermediated settlement: Alice sends 1000 EUR to Bank A. Bank A sends 1000 EUR to Bank B. Bank B sends 1000 EUR to Bob.

We refer to this scenario as *settlement with intermediaries*, or just *intermediated settlement*.

> The Settlement Factory is responsible for generating the correct set of instructions, such that their allocation and approval are effectively possible.

Author: ```suggestion
The Settlement Factory is responsible for generating the correct set of instructions, so that they can be allocated and approved by the respective parties.
```

> The Settlement Factory is responsible for generating the correct set of instructions, such that full allocation and approval are possible.

Reviewer: What does "full allocation and approval" mean in practice?

Reviewer: Ok, I see it in the table below. Maybe add something like "full allocation and approval, as described in the table below"?

Author: What I meant to say is that the instructions can effectively be allocated and approved. This would not be possible if for instance we have an instruction A -> B but these two parties do not have an account at a shared custodian.

Author: Let me know if you like the updated wording and thanks for taking the time to review!

Reviewer: I would maybe prefer a formulation like "so that they can be allocated and approved by the respective parties" (if that is a correct description).

Author: Cool, I like it
